agent/php_stacked_segment.h

```diff
@@ -155,6 +164,386 @@
+ * know no segment caught the exception. We'll apply the exception and
+ * keep/close stacked segments all the way down the stack to clean up dangling
+ * segments. We pop the current segment *C and apply the exception. Because it
+ * has an exception, the segment is kept so we copy the contents of the stacked
+ * segment *C into a segment c we obtained from the slab allocator, and we make
+ * c a child of the stacked segment *B which becomes the current segment.
+ *
+ *  root                        root <                        root
+ *   |                           |                             |
+ *   *A <                        a                             a
+ *    |                          |                             |
+ *    b                          b                             b
+ *    |                          |                             |
+ *    c                          c                             c
+ *
+ * We pop the current segment *B and apply the exception. Because it has an
+ * exception, the segment is kept so we copy the contents of the stacked segment
+ * *B into a segment b we obtained from the slab allocator, and we make b a
+ * child of the stacked segment *A which becomes the current segment.  Then we
+ * pop the current segment *A and apply the exception. Because it has an
+ * exception, the segment is kept so we copy the contents of the stacked segment
+ * *A into a segment a we obtained from the slab allocator, and we make a a
+ * child of the root.  The exception is applied to the root and the txn ends.
+ *
+ * case 4 php_observer_handle_exception_hook - when a new exception is noticed
+ *  root <                      root                      root
+ *                               |                         |
+ *                               *A <                      *A
+ *                                                         |
+ *                                                         *B
+ *
+ * We start out with a root segment, OAPI calls nr_php_observer_fcall_begin(A),
+ * and it starts stacked segment *A and then nr_php_observer_fcall_begin(B)
+ * starts stacked segment *B as a child of *A.
+ *
+ *  root                        root                       root
+ *   |                           |                           |
+ *   *A <                        *A                          *A
+ *                               |                           |
+ *                               *B                          *B <
+ *                               |                           |
+ *                               *C <                        c
+ *
+ * nr_php_observer_fcall_begin(C) starts *C gets started as child
+ * of *B. Function C throws an exception which B catches but 
+ * nr_php_observer_fcall_end(C) is not called so *C remains the 
+ * current segment. B catches the exception and throws another 
+ * exception which triggers the exception hook. At this point we realize the current
```

Comment:
Why isn't the exception hook triggered in case 3, which also deals with an exception that's not caught by anything up to the root? It is a little murky for me when exception hook is called by Zend.

Reply:
The thing to remember about the exception hook is it is called EVERY time an exception occurs but it has no idea if that exception is caught or not so we need to have the extra functionality to check `This` pointers before cleaning up segments.  This specifically calls out that the exception hook is triggered because we want to do any segment cleanup related to the old exception before we record the new exception that was triggered.  In this way we only have to keep track of one exception at a time.

Reply:
Tried to clarify case 3 here: [7be484d](https://github.com/newrelic/newrelic-php-agent/pull/580/commits/7be484d69eb7a10f65828220a8d6a2a27fa9b99b)